File: aletheia-ios-sdk/Sources/AletheiaSDK/AletheiaSDK.swift
import CryptoKit
import Foundation

/// Aletheia iOS SDK — Trusted Capture & Provenance Infrastructure
///
/// Quick start:
/// ```swift
/// import AletheiaSDK
///
/// // Initialize once per app session
/// let sdk = try await AletheiaSDK.initialize(
///     serverURL: URL(string: "https://api.aletheia.io")!,
///     requireBiometric: true
/// )
///
/// // Sign a captured image
/// let result = try await sdk.captureSession.sign(mediaData: imageData)
///
/// // Send bundle to your backend / Aletheia server
/// let response = try await sdk.verification.submit(result.bundle)
/// print(response.trustLevel) // "PROVENANCE_VERIFIED"
/// ```
@MainActor
public final class AletheiaSDK: Sendable {

    // ─── Public surface ───────────────────────────────────────────────────────

    public let captureSession: TrustedCaptureSession
    public let verification:   VerificationClient
    public let attestation:    AttestationClient

    /// SDK version
    public static let version = "2.0.0"

    /// Mandatory scope disclaimer — must be shown to end users
    public static let provenanceScopeNote =
        "Verification confirms provenance integrity, not independent semantic truth of depicted content."

    // ─── Initialization ───────────────────────────────────────────────────────

    private init(session: TrustedCaptureSession, serverURL: URL?) {
        self.captureSession = session
        self.verification   = VerificationClient(serverURL: serverURL)
        self.attestation    = AttestationClient(serverURL: serverURL)
    }

    /// Initialize the Aletheia SDK.
    ///
    /// - Parameters:
    ///   - serverURL:        Base URL of the Aletheia verification server
    ///   - requireBiometric: Require Face ID / Touch ID for signing operations
    ///   - aiDisclosure:     Declare provenance origin of captures
    public static func initialize(
        serverURL:        URL? = nil,
        requireBiometric: Bool = true,
        aiDisclosure:     TrustedCaptureSession.AIDisclosure = .declaredCameraOriginal
    ) async throws -> AletheiaSDK {
        var config             = TrustedCaptureSession.Configuration.default
        config.requireBiometric = requireBiometric
        config.aiDisclosure     = aiDisclosure
        config.serverBaseURL    = serverURL

        let session = try await TrustedCaptureSession.start(config: config)
        return AletheiaSDK(session: session, serverURL: serverURL)
    }

    // ─── Convenience ──────────────────────────────────────────────────────────

    /// Current device ID (derived from Secure Enclave public key)
    public var deviceID: String { captureSession.currentDeviceID }

    /// Current device public key (compressed P-256 point, base64)
    public var publicKey: String { captureSession.currentPublicKey }
}

// ─── Attestation Client ───────────────────────────────────────────────────────

public final class AttestationClient: Sendable {
    private let serverURL: URL?
    init(serverURL: URL?) { self.serverURL = serverURL }

    /// Perform the full App Attest flow and register with the server.
    public func performAttestation(deviceID: String) async throws -> String {
        guard AppAttestManager.isSupported else {
            throw SDKError.appAttestUnsupported
        }

        // 1. Fetch challenge from server
        guard let base = serverURL else {
            throw SDKError.networkError("serverURL not configured")
        }
        var challengeReq = URLRequest(url: base.appendingPathComponent("api/v2/attest/challenge"))
        challengeReq.httpMethod = "POST"
        challengeReq.setValue("application/json", forHTTPHeaderField: "Content-Type")
        challengeReq.httpBody = try JSONEncoder().encode(["device_id": deviceID, "platform": "apple"])

        let (challengeData, _) = try await URLSession.shared.data(for: challengeReq)
        let challengeResp = try JSONDecoder().decode([String: String].self, from: challengeData)
        guard let nonce = challengeResp["challenge_nonce"] else {
            throw SDKError.networkError("Missing challenge_nonce in response")
        }

        // 2. Generate App Attest key
        let keyID = try await AppAttestManager.generateKeyID()

        // 3. Attest with Apple
        let attestationB64 = try await AppAttestManager.attest(keyID: keyID, serverChallenge: nonce)

        // 4. Send to Aletheia server for validation
        var attestReq = URLRequest(url: base.appendingPathComponent("api/v2/attest"))
        attestReq.httpMethod = "POST"
        attestReq.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let body: [String: Any] = [
            "platform": "apple", "device_id": deviceID,
            "attestation_token": attestationB64,
            "challenge_nonce": nonce, "key_id": keyID,
        ]
        attestReq.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (respData, _) = try await URLSession.shared.data(for: attestReq)
        let resp = try JSONDecoder().decode([String: String].self, from: respData)
        return resp["attestation_id"] ?? ""
    }
}

// ─── Verification Client ──────────────────────────────────────────────────────

public final class VerificationClient: Sendable {
    private let serverURL: URL?
    init(serverURL: URL?) { self.serverURL = serverURL }

    public struct VerificationResponse: Codable {
        public let valid:              Bool
        public let trustLevel:         String
        public let trustScore:         Int
        public let hashMatch:          Bool
        public let signatureValid:     Bool
        public let provenanceScopeNote:String
        enum CodingKeys: String, CodingKey {
            case valid, trustLevel = "trust_level", trustScore = "trust_score"
            case hashMatch = "hash_match", signatureValid = "signature_valid"
            case provenanceScopeNote = "provenance_scope_note"
        }
    }

    public func submit(
        bundle:    TrustedCaptureSession.CaptureBundle,
        mediaHash: String
    ) async throws -> VerificationResponse {
        guard let base = serverURL else {
            throw SDKError.networkError("serverURL not configured")
        }
        var req = URLRequest(url: base.appendingPathComponent("api/v2/verify"))
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let body: [String: Any] = [
            "media_hash": mediaHash,
            "bundle": try JSONSerialization.jsonObject(
                with: JSONEncoder().encode(bundle)
            ),
        ]
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, _) = try await URLSession.shared.data(for: req)
        return try JSONDecoder().decode(VerificationResponse.self, from: data)
    }
}

File: aletheia-android-sdk/src/main/kotlin/io/aletheia/sdk/AletheiaSdk.kt
package io.aletheia.sdk

import android.content.Context
import io.aletheia.sdk.capture.HardwareKeyManager
import io.aletheia.sdk.capture.PlayIntegrityManager
import io.aletheia.sdk.capture.TrustedCaptureSession
import io.aletheia.sdk.capture.TrustedCaptureSession.Configuration
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

/**
 * Aletheia Android SDK — Trusted Capture & Provenance Infrastructure
 *
 * Quick start:
 * ```kotlin
 * val sdk = AletheiaSdk.initialize(
 *     context       = applicationContext,
 *     serverBaseUrl = "https://api.aletheia.io",
 * )
 *
 * val result = sdk.captureSession.sign(imageBytes)
 * val response = sdk.submitCapture(result)
 * Log.d("Aletheia", "Trust level: ${response.trustLevel}")
 * ```
 */
class AletheiaSdk private constructor(
    private val context:       Context,
    private val serverBaseUrl: String?,
    val captureSession:        TrustedCaptureSession,
    val keyManager:            HardwareKeyManager,
    val playIntegrity:         PlayIntegrityManager,
) {

    companion object {
        const val VERSION = "2.0.0"

        /** Mandatory disclaimer — must be shown to end users alongside verification results. */
        const val PROVENANCE_SCOPE_NOTE =
            "Verification confirms provenance integrity, not independent semantic truth of depicted content."

        /**
         * Initialize the Aletheia SDK.
         *
         * @param context         Android application context
         * @param serverBaseUrl   Base URL of the Aletheia verification server
         * @param requireBiometric Require biometric authentication for signing
         * @param aiDisclosure    Provenance origin declaration for captures
         */
        fun initialize(
            context:          Context,
            serverBaseUrl:    String? = null,
            requireBiometric: Boolean = true,
            aiDisclosure:     TrustedCaptureSession.AIDisclosure =
                TrustedCaptureSession.AIDisclosure.DECLARED_CAMERA_ORIGINAL,
        ): AletheiaSdk {
            val appContext   = context.applicationContext
            val keyManager   = HardwareKeyManager(appContext)
            val playIntegrity = PlayIntegrityManager(appContext)
            val config       = Configuration(
                requireBiometric = requireBiometric,
                aiDisclosure     = aiDisclosure,
                serverBaseUrl    = serverBaseUrl,
            )
            val session = TrustedCaptureSession.create(appContext, config)
            return AletheiaSdk(appContext, serverBaseUrl, session, keyManager, playIntegrity)
        }
    }

    // ─── Device information ───────────────────────────────────────────────────

    val deviceId:   String get() = captureSession.deviceId
    val publicKey:  String get() = captureSession.publicKey
    val trustLevel: String get() = captureSession.trustLevel

    val isStrongBoxBacked: Boolean get() = keyManager.isStrongBoxAvailable

    // ─── Submit capture to server ─────────────────────────────────────────────

    data class SubmitResponse(
        val valid:              Boolean,
        val trustLevel:         String,
        val trustScore:         Int,
        val hashMatch:          Boolean,
        val signatureValid:     Boolean,
        val provenanceScopeNote:String,
    )

    suspend fun submitCapture(
        result:    TrustedCaptureSession.CaptureResult,
        apiKey:    String? = null,
    ): SubmitResponse = withContext(Dispatchers.IO) {
        val base = serverBaseUrl ?: throw IllegalStateException("serverBaseUrl not configured")
        val body = JSONObject().apply {
            put("media_hash", result.mediaHash)
            put("bundle", JSONObject(result.bundleJson))
        }.toString().toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url("$base/api/v2/verify")
            .post(body)
            .apply { apiKey?.let { header("Authorization", "Bearer $it") } }
            .build()

        val response = OkHttpClient().newCall(request).execute()
        val json = JSONObject(response.body?.string() ?: "{}")

        SubmitResponse(
            valid              = json.optBoolean("valid", false),
            trustLevel         = json.optString("trust_level", "UNTRUSTED_SOURCE"),
            trustScore         = json.optInt("trust_score", 0),
            hashMatch          = json.optBoolean("hash_match", false),
            signatureValid     = json.optBoolean("signature_valid", false),
            provenanceScopeNote = json.optString("provenance_scope_note", PROVENANCE_SCOPE_NOTE),
        )
    }

    // ─── Attestation flow ─────────────────────────────────────────────────────

    suspend fun performAttestation(
        cloudProjectNumber: Long,
        apiKey: String? = null,
    ): String = withContext(Dispatchers.IO) {
        val base = serverBaseUrl ?: throw IllegalStateException("serverBaseUrl not configured")

        // 1. Fetch challenge
        val challengeBody = """{"device_id":"$deviceId","platform":"android"}"""
            .toRequestBody("application/json".toMediaType())
        val challengeResp = OkHttpClient().newCall(
            Request.Builder().url("$base/api/v2/attest/challenge").post(challengeBody).build()
        ).execute()
        val nonce = JSONObject(challengeResp.body?.string() ?: "{}").optString("challenge_nonce")

        // 2. Get Play Integrity token
        val token = playIntegrity.requestIntegrityToken(nonce, cloudProjectNumber)

        // 3. Send to Aletheia server
        val attestBody = JSONObject().apply {
            put("platform", "android")
            put("device_id", deviceId)
            put("attestation_token", token)
            put("challenge_nonce", nonce)
        }.toString().toRequestBody("application/json".toMediaType())

        val attestResp = OkHttpClient().newCall(
            Request.Builder().url("$base/api/v2/attest").post(attestBody)
                .apply { apiKey?.let { header("Authorization", "Bearer $it") } }
                .build()
        ).execute()

        JSONObject(attestResp.body?.string() ?: "{}").optString("attestation_id", "")
    }
}
